### PR TITLE
Refactor and add to navigation and authorization stories

### PR DESCRIPTION
**Pull request by TravisBorgsmiller**
- Refactored and ordered tests (can change according to how we want our test files organized)
- Added base controllers and dashboard controllers for merchants and admins
- reworked sessions new view to use global variables 
- added partials for rending nav bar and changed logic in application controller to render those partials 
   depending on the user
- put dashboard links in namespaces to help with authorization
- added authorization test for different user to restrict access 
- added migration for a one to many relationship with merchants and users 
- added redirect if already logged in functionality
- cleaned up some tests with commented out parts and minor syntax errors to get tests passing

### CODE REVIEW

File: app/controllers/admin/dashboard_controller.rb [hunk 1]
```diff
@@ -0,0 +1,4 @@
+class Admin::DashboardController < Admin::BaseController
+  def index
+  end
```
Comment (cgaddis36) [inline app/controllers/admin/dashboard_controller.rb:3]: Good job implementing baseController, travis

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -1,10 +1,22 @@
...
   def cart
     @cart ||= Cart.new(session[:cart] ||= Hash.new(0))
   end
 
+  def current_user
+    @current_user ||= User.find(session[:user]) if session[:user]
+  end
+
+  def current_merchant
+    current_user && current_user.merchant_employee?
```
Comment (cgaddis36) [inline app/controllers/application_controller.rb:15]: This is exactly what we talked about, looks good

File: app/controllers/sessions_controller.rb [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  def new
+    if current_user
+      login(current_user)
+      flash[:success] = "You are already logged in!"
+    end
+  end
+
+  def create
+    user = User.find_by(email: params[:email])
+    if user && user.authenticate(params[:password])
```
Comment (cgaddis36) [inline app/controllers/sessions_controller.rb:12]: The authenticate is a good addition

File: app/controllers/sessions_controller.rb [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    redirect_to '/'
+  end
+
+  private
+
+    def login(user)
+      if user.default?
+        redirect_to '/profile'
+      elsif user.merchant_employee?
+        redirect_to '/merchant/dashboard'
```
Comment (cgaddis36) [inline app/controllers/sessions_controller.rb:34]: Clean, precise, to the point

**Review by cgaddis36 — APPROVED** [review]

File: app/controllers/application_controller.rb [hunk 5]
```diff
@@ -1,10 +1,22 @@
...
 
-  helper_method :cart
+  helper_method :cart, :current_user, :current_admin, :current_merchant
 
   def cart
     @cart ||= Cart.new(session[:cart] ||= Hash.new(0))
   end
 
+  def current_user
+    @current_user ||= User.find(session[:user]) if session[:user]
```
Comment (adumortier) [inline app/controllers/application_controller.rb:11]: this is what we were missing yesterday, awesome work!

File: app/controllers/application_controller.rb (hunk 2 above)
Comment (adumortier) [reply app/controllers/application_controller.rb:15]: yes checking that the `current_user` exists and only then checking that the `current_user` is indeed a `merchant_employee`

File: app/controllers/cart_controller.rb [hunk 6]
```diff
@@ -1,4 +1,6 @@
 class CartController < ApplicationController
+  before_action :exclude_admin
```
Comment (adumortier) [inline app/controllers/cart_controller.rb:2]: making sure `admin` can not access the cart

File: app/controllers/sessions_controller.rb (hunk 3 above)
Comment (adumortier) [inline app/controllers/sessions_controller.rb:12]: checking that the `password_digest` for the incoming `password` matches the `password_digest` stored in the database

File: app/controllers/sessions_controller.rb [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+      session[:user] = user.id
+      flash[:success] = "You are now logged in!"
+    else
+      flash[:error] = "Incorrect email/password"
+      render :new
+    end
+  end
+
+  def destroy
+    session.clear
```
Comment (adumortier) [inline app/controllers/sessions_controller.rb:23]: clearing the `session` to remove the current user id stored in the `session`

File: app/models/user.rb [hunk 8]
```diff
@@ -0,0 +1,15 @@
+class User < ApplicationRecord
+
+  validates_presence_of :name, :street_address, :city, :state, :zip_code
+  validates_presence_of :password, require: true
+  validates :email, uniqueness: true, presence: true
+
+  validates_confirmation_of :password, :message => "Passwords should match"
+
+  belongs_to :merchant, optional: true
```
Comment (adumortier) [inline app/models/user.rb:9]: the `optional:true` allows to create `users` without them belonging to a `merchant`

File: app/models/user.rb [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+
+  has_secure_password
+
+  enum role: %w(default merchant_employee admin)
```
Comment (adumortier) [inline app/models/user.rb:13]: gives us access to boolean methods on user (user.default? ...) needed for authorization.

File: app/views/layouts/application.html.erb [hunk 10]
```diff
@@ -10,10 +10,17 @@
 
   <body>
     <nav class = "topnav">
-      <%= link_to "All Merchants", "/merchants"%>
-      <%= link_to "All Items", "/items"%>
-      <%= link_to "Cart: #{cart.total_items}", "/cart" %>
-    </nav>
+     <% if current_merchant %>
+       <%= render "partial/merchantnav" %>
```
Comment (adumortier) [inline app/views/layouts/application.html.erb:14]: the `partials` are stored in our `partials` directory

File: app/views/users/new.html.erb [hunk 11]
```diff
@@ -0,0 +1,29 @@
+<h1>New User Form</h1>
+<%= form_tag '/users' do %>
+
+<%= label_tag :name %>
+<%= text_field_tag :name, @user.name %>
+
+<%= label_tag :street_address %>
+<%= text_field_tag :street_address, @user.street_address  %>
```
Comment (adumortier) [inline app/views/users/new.html.erb:8]: looks like a lot nicer than the conditional we had in place here :)

File: config/routes.rb [hunk 12]
```diff
@@ -33,4 +33,21 @@
...
+  end
+
+  get "/register", to: "users#new"
+  post "/users", to: "users#create"
+  get "/profile", to: "users#show"
+
+  get "/", to: "welcome#index"
+  get "/login", to: "sessions#new"
+  post "/login", to: "sessions#create"
+  delete "/logout", to: "sessions#destroy"
```
Comment (adumortier) [inline config/routes.rb:52]: adding `routes` to take care of the `login`/`logout`

File: db/migrate/20200220124550_add_userto_merchants.rb [hunk 13]
```diff
@@ -0,0 +1,5 @@
+class AddUsertoMerchants < ActiveRecord::Migration[5.1]
+  def change
+    add_reference :users, :merchant, foreign_key: true
```
Comment (adumortier) [inline db/migrate/20200220124550_add_userto_merchants.rb:3]: adds a one-to-many relationship between the `Merchant` and `User`

**Review by adumortier — APPROVED** [review]

**Pull request merged**